### PR TITLE
packet: Read BGP peer address from metadata service

### LINKED ISSUE #1009: components/metallb: Ensure correct discovery of BGP peer address
In some Packet facilities the BGP address is the node's gateway address, whereas in others it is a **loopback** address that's reachable via the gateway. Right now the latter case seems to break our MetalLB component.

Right now we do the following to discover the peer address:

https://github.com/kinvolk/lokomotive/blob/3008e5a1cd7cc478482a7c2b365e8a02ae25fc95/assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl#L90

Packet exposes the correct BGP addresses to use generically on their API as well as the metadata service:

```
curl https://api.packet.net/devices/${INSTANCE_UUID}/bgp/neighbors

curl https://api.packet.net/devices/${INSTANCE_UUID} | jq -r .bgp_neighbors[0]

curl https://metadata.packet.net/metadata | jq -r .bgp_neighbors[0]
```

We can likely use the above instead of reading the gateway from the routing table.

### PR DESCRIPTION
**Pull request by johananl**
In some Packet facilities the BGP peer address isn't the same as the gateway address allocated for a host. Rather, it is a loopback address that's reachable via the gateway.

The Packet metadata service now exposes BGP info to hosts, so we can query the metadata service for the BGP peer address.

The source address is explicitly specified since when the peer address is a loopback address, the source IP addresses which ends up getting selected by the kernel is the node's *public* address which doesn't work. In cases where the peer address is the gateway address there is no harm in explicitly specifying the source.

Fixes #1009.

### CODE REVIEW

**Review by invidian — DISMISSED** [review]
LGTM, I wonder if we should also add tests for it, so we can detect some issues in the future. Thanks for quick reaction @johananl !

**Comment by johananl** [comment]
> LGTM, I wonder if we should also add tests for it, so we can detect some issues in the future. Thanks for quick reaction @johananl !

Ideally such a test would be a "blackbox" ingress test (i.e. sending traffic via MetalLB). *However*, there doesn't seem to be a trivial way to handle EIP allocation so we would have to build something for it. Additionally, such a test wouldn't have caught this specific issue because only specific facilities were affected.

**Comment by johananl** [comment]
There is a race condition with enabling BGP.

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (rata) [inline assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:87]: What if curl time out or hangs for a very long time? I think the content of the variable might not be a valid peer and we probably should fail?

I think adding these parameters might help (is not nice, though):  `curl --retry-delay 1 --retry 60 --retry-connrefused --retry-max-time 60 --connect-timeout 20`

(Taken from https://github.com/kinvolk/vpn/blob/923a75d6713690e811fb003a1bf01a9e3398f307/bird/startup.sh#L9, this repo is not public for other seeing it. Sorry :()

That will only help to not hang, but we still need to handle the failure scenario.

Comment (rata) [inline assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:87]: Why the first one of the peers?

**Review by rata — APPROVED** [review]
Added two comments about failure scenarios, but if this is needed ASAP we can handle those in a following PR. As you consider it is best.

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:87]: Good point. We're facing another problem though: sometimes `curl` runs before the BGP session got created by Terraform, in which case the value becomes `null`. I'm trying to avoid having to wrap this in a retry script.

Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:87]: Because MetalLB doesn't support multiple node peers. I've updated the commit message.

Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:99]: I've added a "wait for" script which handles retries. I've added `--connect-timeout` though.

Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:99]: Actually we may want some flags in the actual command which gets the data. Fixing.

**Comment by johananl** [comment]
> There is a race condition with enabling BGP.

Fixed using a retry script.

**Comment by invidian** [comment]
>Additionally, such a test wouldn't have caught this specific issue because only specific facilities were affected.

We randomize the locations to get best machines availability, so it would occur in some cases at least.

>Ideally such a test would be a "blackbox" ingress test (i.e. sending traffic via MetalLB). However, there doesn't seem to be a trivial way to handle EIP allocation (in combination with multiple parallel CI runs) so we would have to build something for it. 

Right... Given that it is not trivial to test, we should merge this to fix issue and perhaps raise the priority of testing it.

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (invidian) [inline assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:448]: What if BGP is not enabled for some reason on the host? Are we able to differentiate somehow situation when it's not enabled completely from not _yet_ enabled?

**Review by invidian — APPROVED** [review]
Just one thought, otherwise LGTM

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:448]: Oh, good catch. We do have conditionals in the kubelet template for when BGP is disabled, however we must also disable this script, otherwise kubelet will never start.

**Comment by johananl** [comment]
Marking as "do not merge" until https://github.com/kinvolk/lokomotive/pull/1010#discussion_r494599881 is resolved.

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (invidian) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:448]: Ah, hehe. I meant more like a bug in Terraform code. I forgot that we actually have an option to disable BGP for workers :)

Comment (rata) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:448]: In case it helps, it is here: https://github.com/kinvolk/lokomotive/blob/master/pkg/platform/packet/template.go#L207

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (rata) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl]: Oh, right, I forgot about that :). Can you create an issue to do this in the future, then? So we don't forget :)

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (rata) [inline assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:458]: Wouldn't it be simpler to write the metadata info to an env file, so that the kubelet just uses that (like Enviromnent=<file created>) instead of repeating the curl?

Furthermore, the following curl can return an error, for example, and in that case we are not handling it correctly

Comment (invidian) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:458]: I guess it make sense, but we need to write it to some tmpfs, so it gets removed on boot, right?

Comment (rata) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:458]: No, if the file is removed the kubelet service won't work after node restart. We want to preserve that file.

Comment (invidian) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:458]: Yeah, but on next boot it will be freshly recreated. Otherwise kubelet will start without waiting for file, so we may get stale data.

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl]: https://github.com/kinvolk/lokomotive/issues/1014

**Comment by johananl** [comment]
@rata @invidian do you guys think the deployment should fail if BGP is enabled bug we couldn't get the peer address? In other words, do we let the kubelet start without the BGP-related labels? Doing so would mean the user will get a working cluster but won't be able to use MetalLB.

**Comment by invidian** [comment]
>@rata @invidian do you guys think the deployment should fail if BGP is enabled but we couldn't get the peer address? In other words, do we let the kubelet start without the BGP-related labels? Doing so would mean the user will get a working cluster but won't be able to use MetalLB.

I think it should fail, yes. I guess that would assume mis-configured project which has BGP disabled?

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:460]: >Otherwise kubelet will start without waiting for file, so we may get stale data.

I don't think there is ever a case where a node's BGP address changes.

**Comment by johananl** [comment]
> > @rata @invidian do you guys think the deployment should fail if BGP is enabled but we couldn't get the peer address? In other words, do we let the kubelet start without the BGP-related labels? Doing so would mean the user will get a working cluster but won't be able to use MetalLB.
> 
> I think it should fail, yes. I guess that would assume mis-configured project which has BGP disabled?

For example. Could also be other things we haven't foreseen. The current behavior is to fail. I agree it's better to fail loudly than quietly here.

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:450]: Added a conditional around enabling `wait-for-bgp.service`.

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:460]: Yes, I've considered that but decided to go with something simpler since the whole thing is a workaround which should hopefully be gone soon. It *is* safer though to go with the file approach. I'll refactor.

Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:456]: Done.

**Comment by johananl** [comment]
I've refactored the script to poll for BGP metadata and write the peer address to `/run/metadata/bgp`. The kubelet unit now uses this file as an environment file when BGP is enabled.

I've tested a manual deployment with BGP both enabled and disabled and this seems to work well.

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (invidian) [inline assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:457]: Hmm, maybe https://stackoverflow.com/a/5284410/2974814 ? :smile:

**Review by invidian — APPROVED** [review]
LGTM

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:457]: Not sure what you're suggesting. Could you explain?

**Comment by johananl** [comment]
@invidian @rata I would like to merge this ASAP if there are no further concerns since we have a broken state in `master` right now. Are there any more blockers? I suggest that for non-blockers we open new issues and iterate later.

File: assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl
Comment (invidian) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:457]: If you insist on validating obtained data, perhaps we should do it right. So using `^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\.|$)){4}$`, as current regex allows values like:
```
256.0.0.0
1.1.1.1.1
```

Which are not correct.

I was laughing, because validating IP addresses with regex is always fun :)

Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl:457]: I don't "insist" on anything, just thought it's better to add at least a minimal validation to catch at least cases where the output is obviously not an IP address. This isn't comprehensive and doesn't handle IPv6 but it's better than nothing.

`1.1.1.1.1` doesn't match on my testing. `256.0.0.0` indeed matches, but so do `0.0.0.0` and `127.0.0.1` which are valid IPs which don't make sense as a peer address...

I'm happy to use a tighter regex if we have one which works reliably.

Comment (johananl) [reply assets/terraform-modules/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl]: I've replaced the regex with `^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\.|$)){4}$` and am testing.

**Review by iaguis — DISMISSED** [review]
lgtm

**Review by invidian — APPROVED** [review]

**Review by invidian — CHANGES_REQUESTED** [review]
Rebase required.

**Comment by johananl** [comment]
> Rebase required.

Thanks.

P.S. GitHub tells me that :wink:

**Comment by invidian** [comment]
>Thanks.

> P.S. GitHub tells me that wink

I know, but it's still showing the PR status in the list as "Waiting for review", which is not really correct :smile:

**Comment by johananl** [comment]
Testing looks good. Need another LGTM.

**Review by iaguis — APPROVED** [review]

**Review by invidian — APPROVED** [review]

**Pull request merged**
